com.spotify/index.js

```diff
@@ -0,0 +1,662 @@
+    return Tp.Helpers.Http.post(url, data, {
+      useOAuth2: this,
+      dataContentType: 'application/json',
+      accept: 'application/json'
+    });
+  }
+
+  // TODO should probably prefix each method with the Spotify API endpoint category it
+  // begins with (i.e. player, library, browse, personalization, etc.)
+  player_play_helper(data = '', options = {
+    useOAuth2: this,
+    dataContentType: 'application/json',
+    accept: 'application/json'
+  }) {
+    console.log("data is " + data)
+    return this.http_put(PLAY_URL, data, options);
+  }
+
+  do_player_pause() {
+    return this.http_put_default_options(PAUSE_URL, '');
+  }
+
+  do_player_play() {
+    console.log("Playing music...");
+    return this.player_play_helper();
+  }
+
+  do_player_next() {
+    return this.http_post_default_options(NEXT_URL, '');
+  }
+
+  do_player_previous() {
+    return this.http_post_default_options(PREVIOUS_URL, '');
+  }
+
+  do_player_shuffle({ shuffle }) {
+    // TODO figure out how to parse shuffle from natural language
+    // and set the state appropriately.
+    // change it to ENUM - then we can hardcode it into the utterance
+    shuffle = shuffle === 'on' ? 'true' : 'false';
+    console.log("shuffle: " + shuffle);
+    let shuffle_url = SHUFFLE_URL + '?state=' + shuffle;
+    return this.http_put_default_options(shuffle_url, '');
+  }
+
+  do_player_repeat({ repeat }) {
+    // TODO figure out how to parse repeat from natural language
+    // and set the state appropriately.
+    // for now, assume it's a string: track, context or off.
+    console.log("repeat: " + repeat);
+    // TODO put shuffle as a query param
+    let repeat_url = REPEAT_URL + '?state=' + repeat;
+    return this.http_put_default_options(repeat_url, '');
+  }
+
+
+
+  /* Search */
+
+  search(query, types){
+      let searchURL = SEARCH_URL + "?q=" + query.toString().replace(' ', '+') + '&type=' + types + '&market=from_token' + '&limit=1';
+      console.log("searching for  " + searchURL);
+      return Tp.Helpers.Http.get(searchURL, {
+          accept: 'application/json',
+          extraHeaders: {
+              'Authorization': 'Bearer ' + this.state.accessToken
+          }
+      }).then(response => {
+          const parsed = JSON.parse(response);
+          console.log(parsed.tracks);
+          return parsed;
+      });
+  }
+
+
+  do_play_song ({toPlay}){   //TODO : DELETE THIS LATER
+    this.search(toPlay,'track').then(searchResults=>{
+        //TODO: Possibly add more error checking here.
+        if (!searchResults.hasOwnProperty('tracks')) return;
+        let data = {'uris': [searchResults.tracks.items[0].uri]};
+        console.log("data is " + JSON.stringify(data));
+        return this.player_play_helper(JSON.stringify(data));
+    });
+  }
+
+  do_play_something({toPlay}){   // TODO : DELETE THIS LATER
+    this.do_play_song({toPlay});
+  }
+
+  do_play_album({toPlay}){
+    this.search(toPlay,'album').then(searchResults=>{
```

Comment:
you must return the promise to have correct error handling (here and in other do functions)